src/librustc/lint/context.rs

```diff
@@ -1082,6 +1089,112 @@ impl LateLintPass for GatherNodeLevels {
     }
 }
 
+enum CheckLintNameResult {
+    Ok,
+    // Lint doesn't exist
+    NoLint,
+    // The lint is either ranamed or removed and a warning was printed
+    Mentioned
+}
+
+/// Checks the name of a lint for its existence, and whether it was
+/// renamed or removed. Prints a warning for renamed and removed
+/// lints. This is over both lint names from attributes and those
+/// passed on the command line. Since it emits non-fatal warnings and
+/// there are *two* lint passes that inspect attributes, this is only
+/// run from the late pass to avoid printing duplicate warnings.
+fn check_lint_name(sess: &Session,
```

Comment:
The logic is just awkward. It's called in two scenarios and some of the errors/warnings are shared, some different. Doing the warning reporting in the caller would introduce duplication.